Sources/TuistCore/Graph/Graph.swift

```diff
@@ -326,7 +326,15 @@ public class Graph: Graphing {
             .map(productDependencyReference)
 
         references.formUnion(otherTargetFrameworks)
+		
+        /// Xcframeworks
+        let xcframeworks = findAll(targetNode: targetNode, test: isXCFramework)
```

Comment:
The issue is with `isDynamicAndLinkable` implementation which looks for if the frameworkNode is of dynamic or static in nature, i am not sure how will the `xcframework` node fare in this case.

Reply:
> That said, to make it work we need to update the xcframework node to provide the correct binary path.

`override var binaryPath: AbsolutePath {
        let frameworkName = "\(path.basenameWithoutExt).framework"
        let filePathsToBinary = [
            "\(path.pathString)/ios-arm64/\(frameworkName)/\(path.basenameWithoutExt)",
            "\(path.pathString)/ios-x86_64-simulator/\(frameworkName)/\(path.basenameWithoutExt)",
            "\(path.pathString)/ios-x86_64-maccatalyst/\(frameworkName)/Versions/A/\(path.basenameWithoutExt)"
        ]
        let binaryAbsolutePaths = filePathsToBinary.map {
            AbsolutePath(stringLiteral: $0)
        }
        _ = try! PrecompiledMetadataProvider().architectures(binaryPath: binaryAbsolutePaths.first!)
        return XCFrameworkNode.binaryPath(frameworkPath: path)
    }`

Reply:
Ya sadly that has the con of hardcoding the paths as you pointed out. Luckily the information we need resides within the `Info.plist` inside the xcframework!

We can beef up the XCFrameworkNode to store more information when we parse/create it

```swift
final class XCFrameworkNode: PrecompiledNode {
    struct Library {
        var path: AbsolutePath
        var architectures: [BinaryArchitecture]
    }
    private let libraries: [Library]
// ...
```

When we parse the XCFramework we can extract this information from the InfoPlist

```swift
static func parse(path: AbsolutePath,
                      cache: GraphLoaderCaching,
                      fileHandler: FileHandling = FileHandler.shared) throws -> XCFrameworkNode {
        if let xcframeworkNode = cache.precompiledNode(path) as? XCFrameworkNode {
            return xcframeworkNode
        }
        let infoPlist = path.appending(component: "Info.plist")
        guard fileHandler.exists(infoPlist) else {
            throw XCFrameworkError.missingInfoPlist(path: infoPlist)
        }
       let data = try fileHandler.read(infoPlist)  // <-- we need a new method on FileHandler that returns `Data` 
        let xcFrameworkInfo = try JSONDecoder().decode(XCFrameworkInfoPlist.self, from: data)
        
        let libraries: [Library] = xcFrameworkInfo.availableLibraries.map {
            let libraryPath = path.appending(component: $0.libraryIdentifier)
                                                   .appending(component: $0.libraryPath)
            let architectures: [BinaryArchitecture] = // ...
            return Library(path: libraryPath,
                           architectures: architectures)
        }
        let xcframeworkNode = XCFrameworkNode(path: path, libraries: libraries)
        
        cache.add(precompiledNode: xcframeworkNode)
        return xcframeworkNode
    }
```

This may also help us later on add some validation later on via linting too (e.g. ensuring the XCFramework you depend on has a valid platform etc... - but we can tackle that separately once we get the basics going)

Reply:
> When we parse the XCFramework we can extract this information from the InfoPlist

I'd not interact with the system as part of the parsing of the model. I think models should just contain a bunch of attributes and helper methods and there should be an utility that given a model, can provide you with extra information about it.

For instance, you can take a look at [this utility](https://github.com/tuist/tuist/blob/master/Sources/TuistCore/Utils/FrameworkMetadataProvider.swift) that Tuist uses to get extra information from a given framework.

Removing side effects from constructors makes things way easier to test.

Reply:
Agreed libraries and architectures info extraction in respective metadata provider will be a suitable place